### ISSUE TITLE
import/no-unresolved errors for .d.ts files since 2.25

### ISSUE THREAD

**Issue by mrdrogdrog** [import-js/eslint-plugin-import#2267]
### Environment

Node: 16.11.1
ESLint: 7.32.0
Operating System: Linux

### What happened?

We tried to update eslint-plugin-import in https://github.com/hedgedoc/react-client from 2.24.2 to 2.25.2 but eslint fails now. It can't resolve paths to `.d.ts` files anymore.

### What did I expect?
No breaking changes in behaviour since it isn't a major release.

### Additional Information

Feel free to take a look at our CI log.
https://github.com/hedgedoc/react-client/runs/3915016791?check_suite_focus=true

Thanks for your help

**Comment by andersk** [comment]
This was introduced by 4d15e268b694e9e9946ea16d9e1b94de1b850d7c (#2220). Cc @jabiko.

Minimal reproduction:

```ts
// foo.ts
import { bar } from "./bar";
console.log(bar);
```

```ts
// bar.d.ts
export const bar: string;
```

```json
// .eslintrc.json
{
  "plugins": ["@typescript-eslint", "import"],
  "parser": "@typescript-eslint/parser",
  "extends": "plugin:import/typescript",
  "rules": {
    "import/no-unresolved": "error"
  }
}
```

```console
$ npm i -D @typescript-eslint/eslint-plugin@5.0.0 @typescript-eslint/parser@5.0.0 eslint@7.32.0 eslint-plugin-import@2.25.2 typescript@4.4.4
$ npx eslint foo.ts
 
/home/anders/zulip/test/eslint-import-bug/foo.ts
  2:21  error  Unable to resolve path to module './bar'  import/no-unresolved

✖ 1 problem (1 error, 0 warnings)
```

**Comment by mrdrogdrog** [comment]
In a pure typescript environment it's normal to use .d.ts files that just provide types or interfaces but no functions. E.g. if you want to share types between multiple ts files.

**Comment by mrdrogdrog** [comment]
Just tested: One way to make this work without renaming all `.d.ts` files to `.ts` is to replace the import.

`import { bar } from "./bar";`
becomes
`import type { bar } from "./bar";`

A bit annoying if you have many many .d.ts files like
![image](https://user-images.githubusercontent.com/6124140/137633179-be8095b1-d52a-4ec4-b4f0-5b065107d3c0.png)

**Comment by ljharb** [comment]
@mrdrogdrog it's *not* normal to have a .d.ts file and no corresponding .js file to be imported.

Yes, breaking changes in patch releases are bad, but every single bug fix is a breaking change if someone's relying on the bug.

I see what you mean tho - in this case, you're importing a type, but you weren't using the preferred `import type` syntax for it, so it shouldn't have warned but did.

I think the issue is that you're not using the TS resolver (but neither is https://github.com/import-js/eslint-plugin-import/blob/main/config/typescript.js#L18-L22).

If you add the TS resolver - above `node` - does it work?

**Comment by mrdrogdrog** [comment]
I read the typescript and eslint docs again and changed my code to use "import type" and the "consitent-type-import" rule.

Maybe you should improve the error message in this case? Because if you're used to "this import should work because it always did" then "unable to resolve module" is a bit confusing :/

**Comment by Ulrikop** [comment]
I get also problems with `.d.ts` for `import/extensions` now. I believe it belongs to the problems described here, so I don' create a new bug for it.

I'm using the config `'import/extensions': ['error', 'always', { pattern: { js: 'never', ts: 'never' }, ignorePackages: true }]`.

If I have a file `foo.d.ts` without a corresponding `.js` file, I'm getting now errors for import like `from './foo'`.

> @mrdrogdrog it's not normal to have a .d.ts file and no corresponding .js file to be imported.

That is wrong in my opinion. e.g. when I want to describe the expected messages of a communication I add them to an own file.
Or if I describe an option param of a method which is used in different files (because it is passed through or so) I want to have it also described in an own `options.d.s` file.

So it could be, that the most are not doing this but I want to be able to do it like I think it is good. Can we provide an `option` to that rules to get the old behavior again?

But for now I have to disable all the rules :-(

**Comment by jablko** [comment]
@mrdrogdrog :+1:

@Ulrikop Thinking more about this, and TypeScript maintainers amenity to value usage of type-only const enums: https://github.com/microsoft/TypeScript/issues/42991#issue-817625053

*If* isolatedModules matters to you, could you rename `foo.d.ts` -> `foo.ts`? That would simultaneously pacify eslint-plugin-import. The compiler will elide `import { Foo } from './foo'` if isolatedModules is false, and preserve it if true.

**Comment by ljharb** [comment]
@Oliver84 this issue is about d.ts files, which only apply if you’re using Typescript. Can you file a new issue if you’re seeing something unrelated to TS?

**Comment by Ulrikop** [comment]
Sorry for my late answer. due to an emergency in my family, I don't have time to test anything at the moment (e.g. adding `.d.ts` to config).
Even though I think that having your own rule to check whether `.d.ts` files are only imported as `type` is better, so as not to force anyone who wants to use the other rules to do so,I will find a way so that my company can use the eslint rules (patch the old behavior, find a way to configure it to work for us, change the code, ... )

**Comment by JounQin** [comment]
Have you tried [`eslint-import-resolver-typescript`](https://github.com/alexgorbatchev/eslint-import-resolver-typescript)?

**Comment by jablko** [comment]
@ZJS248 Thanks for the code, I haven't tried linting it myself yet ... Can you elaborate on the false positives/it causes errors that you're seeing?

My belief is that the import/no-unresolved rule currently skips `import type` declarations, so either you're seeing other errors, or my belief is wrong?

**Comment by ZJS248** [comment]
@jablk
errors like this 
include `import/named`  `import/no-extraneous-dependencies`  `import/no-cycle` etc.
![screenshot](https://user-images.githubusercontent.com/32304065/144339740-49e5430f-167f-48ad-a9e7-0fe1a1973744.png)
o

**Comment by JounQin** [comment]
You should use https://github.com/import-js/eslint-import-resolver-typescript
